src/main/java/edu/kis/powp/jobs2d/command/visitor/CanvasBoundaryCheckVisitor.java

```diff
@@ -0,0 +1,59 @@
+package edu.kis.powp.jobs2d.command.visitor;
+
+import edu.kis.powp.jobs2d.command.ComplexCommand;
+import edu.kis.powp.jobs2d.command.OperateToCommand;
+import edu.kis.powp.jobs2d.command.SetPositionCommand;
+import java.awt.Dimension;
+import java.awt.Point;
+
+public class CanvasBoundaryCheckVisitor implements CommandVisitor {
+    private int minX = Integer.MAX_VALUE;
+    private int maxX = Integer.MIN_VALUE;
+    private int minY = Integer.MAX_VALUE;
+    private int maxY = Integer.MIN_VALUE;
+
+    private boolean exceedsCanvas = false;
+
+    private final Dimension canvasDimension;
+
+    public CanvasBoundaryCheckVisitor(Dimension canvasDimension) {
+        this.canvasDimension = canvasDimension;
+    }
+
+    @Override
+    public void visitOperateToCommand(OperateToCommand operateToCommand) {
+        updateDrawingBoundaryLimits(operateToCommand.getPosX(), operateToCommand.getPosY());
+    }
+
+    @Override
+    public void visitSetPositionCommand(SetPositionCommand setPositionCommand) {
+        updateDrawingBoundaryLimits(setPositionCommand.getPosX(), setPositionCommand.getPosY());
+    }
+
+    @Override
+    public void visitComplexCommand(ComplexCommand complexCommand) {
+        complexCommand.iterator().forEachRemaining(command -> command.accept(this));
+        if (doesExceedCanvas()) {
+            System.err.println("Drawing exceeds the canvas boundaries.");
+        }
+    }
+
+    private void updateDrawingBoundaryLimits(int x, int y) {
+        minX = Math.min(minX, x);
+        maxX = Math.max(maxX, x);
+        minY = Math.min(minY, y);
+        maxY = Math.max(maxY, y);
+        checkCanvasBounds();
+    }
+    private void checkCanvasBounds() {
+        Point topLeft = new Point(-canvasDimension.width / 2, -canvasDimension.height / 2);
+        Point bottomRight = new Point(canvasDimension.width / 2, canvasDimension.height / 2);
+        if(minX < topLeft.x || maxX > bottomRight.x || minY < topLeft.y || maxY > bottomRight.y) {
+            exceedsCanvas = true;
```

Comment:
Not the best solution. This is not the stateless approach. Once exceedsCanvas is set to true, then visitor becomes unusable. The best way, would be to set exceedsCanvas accordingly to the outer accept call.